### ISSUE TITLE
MS Edge Issue: Rendering all vector fills as black instead of appropriate color

### ISSUE THREAD

**Issue by joshreed** [canvg/canvg#520]
I saw issue #416 (and several others). The fixes suggested there do in fact help, however it does not completely solve the problem in Microsoft Edge 38.14393.0.0. I certainly ensured that the patch was applied. I was able to "work around the problem" after finding a suggestion here on Stack Overflow: http://stackoverflow.com/a/38848733

For reference for others, I was able to "fix" it by dynamically removing the "xmlns" attribute right before passing the SVG to canvg (specifically targeting this removal for Edge's user agent only). Not sure that this is the greatest solution, but it at least works for my needs at this point. 

There may be a way to fix it in the canvg code itself, especially for those that are more familiar with the canvg codebase itself. I gave it a try, but everywhere I tried to put in the removal of the xmlns attribute in the codebase it was "too late" and didn't affect the final canvas renderings. In the interest of my deadline, I gave up and just handled it before passing the SVG to canvg. But I would certainly love to see a more permanent solution, if possible.

**Comment by DRippstein** [comment]
@joshreed: Can you clarify what the problem is? It's hard to accept your solution when you haven't told us the problem it's trying to solve. Example code would be nice too.

I'm also using Microsoft Edge 38.14393.0.0 and combined with the latest version of CanvasVG, I'm able to see vector fills rendered just fine visually.

Now, if I use a pre-#416 version of CanvasVG, I see black vectors... and applying your solution certainly works to trick Edge into rendering fills correctly... but it seems unnecessary in the current version given that #416 works just fine for others including myself (unless you provide some scenario we haven't considered).

**Comment by joshreed** [comment]
Well perhaps I am not using the latest release then? I had to manually apply the fix provided in https://github.com/canvg/canvg/pull/441/files and I don't see any Tags or Branches with newer commits than the date that that PR was created. And as I mentioned above, that fix did help, but didn't fully solve my problem. My issue was occurring on this image

````
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 37.7 47.7" enable-background="new 0 0 37.7 47.7" xml:space="preserve">
<g id="tower">
	<g>
		<g>
			<path fill="#404563" d="M19,1.2c-9.3,0-17.9,7-17.9,16.5C1.2,29.5,19,46.2,19,46.2s17.9-16.7,17.9-28.5C36.9,8.2,28.4,1.2,19,1.2
				z"/>
		</g>
	</g>
	<g>
		<g>
			<path fill="#FFFFFF" d="M10.3,30.1c0.6,0,1.8-2.7,2-3.2c0.7-1.4,1.4-2.8,1.9-4.2c1.6-4.1,2.5-8.3,3.4-12.5c0-0.1,0-0.1,0-0.2h0.9
				c0,0-1.9,9.9-2.3,11.1s-2.5,9.1-2.5,9.1S10.3,30.1,10.3,30.1z"/>
			<path fill="#FFFFFF" d="M27.2,30c-0.6,0-1.8-2.7-2-3.2c-0.7-1.4-1.4-2.8-1.9-4.2c-1.6-4.1-2.5-8.3-3.4-12.5c0-0.1,0-0.1,0-0.2
				h-0.9c0,0,1.9,9.9,2.3,11.1c0.4,1.2,2.5,9.1,2.5,9.1S27.2,30,27.2,30z"/>
			<rect x="17.8" y="9.9" fill="#FFFFFF" width="1.9" height="20.2"/>
			<rect x="15.8" y="20.3" fill="#FFFFFF" width="5.8" height="0.9"/>
			<rect x="14.7" y="24.4" fill="#FFFFFF" width="8.4" height="0.9"/>
		</g>
		<circle fill="#FFFFFF" cx="18.8" cy="8.1" r="2.9"/>
	</g>
</g>
</svg>
````

Additionally, paste that into http://canvg.github.io/canvg/examples/index.htm to see the problem. Similarly the "pattern" and "transform" examples (among others) render black in MS Edge.

**Comment by yasirrose** [comment]
I am also facing the same issue. Any update on it. it render black in MS Edge.

**Comment by zeronius** [comment]
Same issue - when i have e.g. polygon element with fill attribute, the shape is rendered as black in Edge.
`<svg id="svg-scene" xmlns="http://www.w3.org/2000/svg" width="1000" height="500" viewBox="0 0 1920 1080">
      <polygon points="486.5,420.5 1643,230  1648.5,958.5 1284.5,885 1283,400 615,463 619,713.5 561.5,724.5 486.5,704"
        fill="rgb(255,0,0)"/>
    </svg>`

Also it is rendered when opacity is 0:

` <svg id="svg-scene" xmlns="http://www.w3.org/2000/svg" width="1000" height="500" viewBox="0 0 1920 1080">
      <polygon points="486.5,420.5 1643,230  1648.5,958.5 1284.5,885 1283,400 615,463 619,713.5 561.5,724.5 486.5,704"
        fill="rgb(255,0,0)" opacity="0"/>
    </svg>`

**Comment by karthickDhanapal** [comment]
Is there any fix available

**Comment by joshreed** [comment]
Honestly, I have not checked to see if a fix was actually made in the codebase. However, I was able to work around the issue by setting up a proxy that deleted the `xmlns` attribute on the SVG file that was ultimately served up (only used when the browser in use is MS Edge).

Basically the JavaScript side would do this:
```
// Special for Microsoft Edge, proxy the SVGs to remove the 'xmlns' attribute before passing to canvg...why this works...idk
if(/Edge/.test(navigator.userAgent)){
   canvg('canvg_'+canv_i, '/proxy.svg?filename='+encodeURI($(this).data('file-name')));
} else {
   canvg('canvg_'+canv_i, $(this).attr('src'));
}
```

And then I had Ruby (Rails) code that looked like this. You may have to translate into a different language, depending on your use-case:
```
  def svg_no_xmlns
    image_param = params[:filename]
    image_file = Rails.root.join('app', 'assets', 'images', image_param)
    begin
      file_contents = open(image_file) { |f| f.read }
    rescue
      Rails.logger.warn "File not able to be opened, opening default file."
      file_contents = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n<!-- Generator: Adobe Illustrator 18.1.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->\r\n<svg version=\"1.1\" id=\"Layer_1\"  xmlns:xlink=\"http://www.w3.org/1999/xlink\" x=\"0px\" y=\"0px\"\r\n\t viewBox=\"0 0 41.2 50\" enable-background=\"new 0 0 41.2 50\" xml:space=\"preserve\">\r\n<g id=\"building\">\r\n\t<g>\r\n\t\t<g>\r\n\t\t\t<path fill=\"#80449f\" d=\"M20.7,0C10.3,0,0.9,7.8,0.9,18.4C0.9,31.5,20.7,50,20.7,50s19.8-18.5,19.8-31.6C40.5,7.8,31.1,0,20.7,0z\"\r\n\t\t\t\t/>\r\n\t\t</g>\r\n\t</g>\r\n\t<path fill=\"#FFFFFF\" d=\"M24.2,12.2V8.6h-6.9v7.5h-4.4v15.1h15.6v-19H24.2z M17.3,28.6h-2v-3.1h2V28.6z M17.3,24.9h-2v-3.1h2V24.9z\r\n\t\t M17.3,21.1h-2V18h2V21.1z M21.7,28.6h-2v-3.1h2V28.6z M21.7,24.9h-2v-3.1h2V24.9z M21.7,21.1h-2V18h2V21.1z M21.7,17.3h-2v-3.1h2\r\n\t\tV17.3z M21.7,13.6h-2v-3.1h2V13.6z M26.1,28.6h-2v-3.1h2V28.6z M26.1,24.9h-2v-3.1h2V24.9z M26.1,21.1h-2V18h2V21.1z M26.1,17.3h-2\r\n\t\tv-3.1h2V17.3z\"/>\r\n</g>\r\n</svg>\r\n"
    end
    send_data file_contents.gsub(/xmlns=.*\/2000\/svg"/, ''), :type => 'image/svg+xml'
  end
```

In short, it takes a filename as a parameter. It opens up the file (text), and then does a Regex substitution to delete the `xmlns` attribute, and then sends the resulting data back to the browser with the `content-type` as `image/svg+xml` so that the browser will still treat it as an SVG.

Hopefully this is helpful. It's a pain, but it works.

**Comment by gabelerner** [comment]
i tested in ie11 it seems to work, has edge fixed this yet? please reopen if not.
